rasterio/mask.py

```diff
@@ -73,37 +75,38 @@ def mask(raster, shapes, nodata=None, crop=False, all_touched=False,
         rasterio.features.bounds(shape, north_up=north_up) for shape in shapes]
     lefts, bottoms, rights, tops = zip(*all_bounds)
 
-    if pad:
-        dx = raster.res[0] / 2
-        dy = raster.res[1] / 2
-    else:
-        dx = 0.0
-        dy = 0.0
-
-    if north_up:
-        mask_bounds = (min(lefts) - dx, min(bottoms) - dy,
-                       max(rights) + dx, max(tops) + dy)
-    else:
-        mask_bounds = (min(lefts) - dx, max(bottoms) + dy,
-                       max(rights) + dx, min(tops) - dy)
-
     source_bounds = raster.bounds
 
-    # Raise or warn about bounds mismatches.
-    if rasterio.coords.disjoint_bounds(source_bounds, mask_bounds):
-        if crop:
-            raise ValueError("Input shapes do not overlap raster.")
+    if crop:
+        if pad:
+            dx = raster.res[0] / 2
+            dy = abs(raster.res[1] / 2)
+        else:
+            dx = 0.0
+            dy = 0.0
+
+        if north_up:
+            mask_bounds = (min(lefts) - dx, min(bottoms) - dy,
+                           max(rights) + dx, max(tops) + dy)
         else:
-            warnings.warn("GeoJSON outside bounds of existing output " +
-                          "raster. Are they in different coordinate " +
-                          "reference systems?")
+            mask_bounds = (min(lefts) - dx, max(bottoms) + dy,
+                           max(rights) + dx, min(tops) - dy)
+
+        # Raise or warn about bounds mismatches.
+        if rasterio.coords.disjoint_bounds(source_bounds, mask_bounds):
+            if crop:
+                raise ValueError("Input shapes do not overlap raster.")
+            else:
+                warnings.warn("GeoJSON outside bounds of existing output " +
+                              "raster. Are they in different coordinate " +
+                              "reference systems?")
 
-    if crop:
         bounds_window = raster.window(*mask_bounds)
```

Comment:
This seems to be where we are introducing a potential bug.  Because we are bringing in a small epsilon value when evaluating the window, the window we get back is just slightly different than we pass in.  (or I'm just misunderstanding what it is supposed to be doing)

Reply:
@brendan-ward right. I think it would be best to avoid calling ``.window()`` if we can avoid it. If pad is `False` and `dx = dy = 0.0`, we could set `bounds_window = Window(0, 0, raster.width, raster.height)`. Does that make sense to you?

Reply:
Agreed, we can simplify if `pad` is `False`.  However, the issue only presents when`pad` is `True` so that's where we need to focus more of our efforts.

Reply:
I looked at this further.  We still need to call `.window()` if `pad` is `False` because `crop` is `True`, and we are just reading out the overlapping region of data from the underlying raster.